New release api

SuperArnold opened the pull request:
Add crawler_get_new_release.py to get new release album and these
album's songs.
Add Spotify_Parameters Client_ID and Client_Secret of the config.cfg to
update

SuperArnold committed:
New realease api

Add crawler_get_new_release.py to get new release album and these
album's songs.
Add Spotify_Parameters Client_ID and Client_Secret of the config.cfg to
update

AbnerLin commented on daemon/authorization/get_album_song.py:
```diff
@@ -0,0 +1,43 @@
+import urllib.request
+import json
+
+class Get_album_song:
```
According PEP8, Class names should normally use the CapWords convention.

AbnerLin commented on daemon/authorization/crawler_get_new_release.py:
```diff
@@ -0,0 +1,35 @@
+import configparser
```
I think this script should be modular.

rockers7414 commented on daemon/config.cfg:
```diff
@@ -12,3 +12,7 @@ level=DEBUG
+
+[Spotify_Parameters]
+Client_ID=e9676baa69da445bbd9b691658628b69
+Client_Secret=20727079480e4093a14529aefd950ffa
```
I think this id and secret is not proper in the config  file unless they're for development only. So please remove this sensitive data from the config file. Or just don't commit this settings in the branch.

SuperArnold replied:
I have to use id and secret to get token. Can I putting these in my source code?

rockers7414 replied:
Sure. But just don't commit to the remote branch.

rockers7414 commented on daemon/config.cfg:
```diff
@@ -12,3 +12,7 @@ level=DEBUG
 host=0.0.0.0
 port=50000
+
+[Spotify_Parameters]
```
Section header should be uppercase.

rockers7414 commented on daemon/authorization/get_album_song.py:
```diff
@@ -0,0 +1,43 @@
+import urllib.request
```
Please change to use pycurl since the pycurl is implemented by C code. The speed is more faster than request.

rockers7414 commented on daemon/authorization/get_album_song.py:
```diff
@@ -0,0 +1,43 @@
+            "Authorization": "Bearer " + access_token
+        } 
+# }
+    def get_album_song(self,album_id):
```
I think this function is duplicated with getTracksByAlbumId in the class of SpotifyProvider.

rockers7414 commented on daemon/authorization/get_new_release.py:
```diff
@@ -0,0 +1,45 @@
+            "Authorization": "Bearer " + access_token
+        } 
+# }
+    def get_new_list(self,count):
```
I think you should add this function into the interface of MusicVideoInfoProvider and implement this function in SpotifyProvider.

rockers7414 reviewed (changes requested):
Please revise pre request.

cclien0725 commented on daemon/authorization/crawler_get_new_release.py:
```diff
@@ -0,0 +1,35 @@
+
+new_release = Get_new_release(access_token)
+album_song = Get_album_song(access_token)
+
```
I think the two class(`Get_new_release` and `Get_album_song`) you defined maybe can merge into one class or write down these methods into `daemon/provider/musicvideoinfo/spotifyprovider.py`, because it provides Spotify's functionalities.

BTW, the naming rule of classes should be nouns and the coding convention must follow PEP8.

cclien0725 commented on daemon/authorization/crawler_get_new_release.py:
```diff
@@ -0,0 +1,35 @@
+
+# Get new release name. 
+# You will get twenty times as number
+new_count = 1
```
I don't understand what the variable `new_count` is mean, and why must assign to `1`.

cclien0725 commented on daemon/authorization/crawler_get_new_release.py:
```diff
@@ -0,0 +1,35 @@
+# You will get twenty times as number
+new_count = 1
+album_information_list = []
+new_list = new_release.get_new_list(new_count)
```
Maybe you must comment down what the structure is in each element on `new_list`.
It may cause some confusion.

cclien0725 commented on daemon/authorization/crawler_get_new_release.py:
```diff
@@ -0,0 +1,35 @@
+new_release_imformation = []
+# album_information_list is new release album and album's songs
+while new_list_count < len(new_list):
+    album_information_list.append(album_song.get_album_song(new_list[new_list_count]["id"]))
```
I think the inline code must split into two or more line for readability. Because it invoked method, access array, and object, it makes more complexity.

cclien0725 commented on daemon/authorization/get_album_song.py:
```diff
@@ -0,0 +1,43 @@
+        d1 =  json.loads(contentBytes )
+
+        for i,t in enumerate(d1["tracks"]["items"]):
+            json_value = {"name :" : t["name"],
```
Why did it append ` :` in each key on the dictionary type of `json_value`? 
- ex. `"name :"`

Does it for some data processing?

cclien0725 reviewed (changes requested):
There are some suggestions for you, please revise it. Thanks.

SuperArnold committed:
Add new release

The "pass" modified to "raise".
Add function to fetch new release from Spotify.

SuperArnold committed:
Merge branch 'master' into new-release-Arnold

rockers7414 commented on daemon/provider/musicvideoinfo/spotifyprovider.py:
```diff
@@ -173,3 +174,46 @@ def get_tracks_by_album_id(self, album_id):
+        """
+            Get client_id and client_secret value from local.
+        """
+        config = configparser.ConfigParser()
```
About config file, I think you should use the `config.py`.  You can refer the `xmusic.py`.

rockers7414 commented on daemon/provider/musicvideoinfo/spotifyprovider.py:
```diff
@@ -173,3 +174,46 @@ def get_tracks_by_album_id(self, album_id):
+            client.setopt(pycurl.URL, service_host + "/v1/browse/new-releases" + "?" + urlencode(params))
+            client.setopt(pycurl.WRITEFUNCTION, buf.write)
+#                 SSL certificate for Windows
+#            client.setopt(pycurl.CAINFO, "C:\python3\curl\ca-bundle.crt")
```
If this is not needed, please remove this comment.

rockers7414 reviewed (changes requested):
Please help to revise

cclien0725 commented on daemon/provider/musicvideoinfo/musicvideoinfoprovider.py:
```diff
@@ -13,3 +13,6 @@ def get_albums_by_artist_id(artist_id):
     def get_tracks_by_album_id(album_id):
         raise NotImplementedError("get_tracks_by_album_id is not implemented.")
+
+    def get_new_release_by_album_id(offset_count,limit):
```
The interface name: `get_new_release_by_album_id` is not equal to the implemeted name: `getNewRelease`.

rockers7414 replied:
We should get all new release albums one time. Hence, we don't need to give the `offset` and the `limit`.
So you should refactor this function
```
def get_new_release_albums():
    raise NotImplementedError("get_new_release_albums is not implemented.")
```

cclien0725 commented on daemon/provider/musicvideoinfo/spotifyprovider.py:
```diff
@@ -173,3 +174,46 @@ def get_tracks_by_album_id(self, album_id):
 
         return tracks
+
+	def getNewRelease(self, offset_count, limit):
```
The implemented name: `getNewRelease` is not equal to the interface name: `get_new_release_by_album_id`.

cclien0725 reviewed (changes requested):
Please revise these comments. Thanks.

SuperArnold committed:
Update rocker and cclien requests

Change def name
Get cleint_id and client_secret value from config.py

rockers7414 reviewed (commented):
I am a little confused. I think the method will get the album's information and the artist's information. But the data which you return only has artist information.

SuperArnold committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into new-release-Arnold

SuperArnold committed:
Update new release

Add artist information

rockers7414 commented on daemon/provider/musicvideoinfo/spotifyprovider.py:
```diff
@@ -173,3 +177,49 @@ def get_tracks_by_album_id(self, album_id):
 
         return tracks
+
+    def get_new_release_by_album_id(self, offset_count, limit):
```
According to the above changed request, this function should be renamed to `get_new_release_albums()`. And the expected result should be the list of the album.

rockers7414 reviewed (changes requested):
I think you're almost there. Great job! Have a happy coding.

SuperArnold committed:
Update new release

Fetch all new release album

rockers7414 commented on daemon/provider/musicvideoinfo/musicvideoinfoprovider.py:
```diff
@@ -13,3 +13,6 @@ def get_albums_by_artist_id(artist_id):
     def get_tracks_by_album_id(album_id):
         raise NotImplementedError("get_tracks_by_album_id is not implemented.")
+
+    def get_new_release_by_album_id():
```
The function name should change to `get_new_release_albums` since we don't give the album id to this function.

rockers7414 commented on daemon/provider/musicvideoinfo/spotifyprovider.py:
```diff
@@ -173,3 +177,54 @@ def get_tracks_by_album_id(self, album_id):
+            Get client_id and client_secret value from local.
+        """
+        config = Config("../../xmusic.cfg")
+        print("1")
```
Please remove unused testing code.

rockers7414 commented on daemon/provider/musicvideoinfo/spotifyprovider.py:
```diff
@@ -173,3 +177,54 @@ def get_tracks_by_album_id(self, album_id):
 
         return tracks
+
+    def get_new_release_by_album_id(self):
```
Remember to follow the interface to implement.

rockers7414 commented on daemon/provider/musicvideoinfo/spotifyprovider.py:
```diff
@@ -173,3 +177,54 @@ def get_tracks_by_album_id(self, album_id):
+            else:
+                break
+
+        return new_release
```
It's much better to get `Album[]` which is defined in the `entity.py` as the result.

SuperArnold committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into new-release-Arnold

# Conflicts:
#	daemon/provider/musicvideoinfo/musicvideoinfoprovider.py

rockers7414 committed:
Fix some bugs and refactor the new release api.

cclien0725 reviewed (approved):
Looks good, these changes will be merged into master, great job!
